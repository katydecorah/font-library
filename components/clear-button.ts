class ClearButton extends HTMLElement {
  constructor() {
    super();
  }
  connectedCallback() {
    // add aria-role
    this.setAttribute("role", "button");
    this.addEventListener("click", () => {
      this.dispatchEvent(
        new CustomEvent("clear-filter", {
          bubbles: true,
          composed: true,
          detail: {
            filter: this.getAttribute("value"),
          },
        })
      );
    });
  }
}

customElements.define("clear-button", ClearButton);
