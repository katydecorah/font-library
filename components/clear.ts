class ClearButton extends HTMLElement {
  constructor() {
    super();
  }
  connectedCallback() {
    this.addEventListener("click", () => {
      this.dispatchEvent(
        new CustomEvent("clear-filter", {
          bubbles: true,
          composed: true,
        })
      );
    });
  }
}

customElements.define("clear-button", ClearButton);
