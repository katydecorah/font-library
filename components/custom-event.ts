export default function customEvent(
  eventName: string,
  detail?: {
    value?: string | boolean | number;
    id?: string;
  }
): CustomEvent {
  return new CustomEvent(eventName, {
    bubbles: true,
    composed: true,
    detail,
  });
}
