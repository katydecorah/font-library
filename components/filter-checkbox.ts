import customEvent from "./custom-event";

class FilterCheckbox extends HTMLInputElement {
  constructor() {
    super();
    this.addEventListener("change", this.onChange);
    this.handleInitialValue();

    // Listen for events to clear filter
    window.addEventListener("remove-checkbox", () => {
      this.checked = false;
      this.onChange();
    });
  }

  onChange() {
    this.dispatchEvent(
      customEvent("handle-filter", {
        value: this.checked,
        id: this.id,
      })
    );
    this.setUrlParam();
  }

  setUrlParam() {
    const urlParameters = new URLSearchParams(window.location.search);
    // only set variable if it's true
    if (this.checked === false) {
      urlParameters.delete("variable");
    } else {
      urlParameters.set("variable", "true");
    }
    window.history.replaceState(
      {},
      "",
      `${window.location.pathname}?${urlParameters.toString()}`
    );
  }

  handleInitialValue() {
    const urlParameters = new URLSearchParams(window.location.search);
    const initialValue = urlParameters.get("variable");
    if (initialValue === "true") {
      this.checked = true;

      // Wait for main-app to load before dispatching event
      window.addEventListener("main-app-loaded", () => {
        this.onChange();
      });
    }
  }
}

customElements.define("filter-checkbox", FilterCheckbox, { extends: "input" });
