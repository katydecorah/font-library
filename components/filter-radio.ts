class FilterRadio extends HTMLInputElement {
  constructor() {
    super();
    this.onchange = this.onChange;

    // Get init value from URL param
    const urlParams = new URLSearchParams(window.location.search);
    const tag = urlParams.get("tag");
    if (tag === this.value) {
      this.checked = true;
    }
  }

  onChange() {
    this.dispatchEvent(
      new CustomEvent("tag-button-selected", {
        bubbles: true,
        composed: true,
        detail: {
          value: this.value,
        },
      })
    );

    this.setUrlParam();
  }

  setUrlParam() {
    const urlParams = new URLSearchParams(window.location.search);
    urlParams.set("tag", this.value.toString());
    window.history.replaceState(
      {},
      "",
      `${window.location.pathname}?${urlParams.toString()}`
    );
  }
}

customElements.define("filter-radio", FilterRadio, { extends: "input" });
