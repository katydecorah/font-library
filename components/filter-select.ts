class FilterSelect extends HTMLSelectElement {
  constructor() {
    super();
    this.onchange = this.onChange;

    // Get initial value from URL
    const urlParams = new URLSearchParams(window.location.search);
    const initValue = urlParams.get(this.dataset.param);
    if (initValue) {
      if (!this.options.namedItem(initValue)) return;
      this.value = initValue;

      // Wait for main-app to load before dispatching event
      window.addEventListener("main-app-loaded", () => {
        this.onChange();
      });
    }

    // Listen for events to clear filter
    window.addEventListener("remove-select", (e: CustomEvent) => {
      if (e.detail.filter === this.id) {
        this.value = "";
        this.onChange();
      }
    });

    // Listen for changes by other tag elements (radio, button)
    window.addEventListener("tag-button-selected", (e: CustomEvent) => {
      if (this.id === "selectedTag" && this.value !== e.detail.value) {
        this.value = e.detail.value;
      }
    });
  }

  onChange() {
    const eventName =
      this.id === "selectedTag" ? "tag-button-selected" : "handle-filter";
    this.dispatchEvent(
      new CustomEvent(eventName, {
        bubbles: true,
        composed: true,
        detail: { id: this.id, value: this.value },
      })
    );

    this.setUrlParam();
  }

  setUrlParam() {
    const param = this.dataset.param;
    const urlParams = new URLSearchParams(window.location.search);
    if (!this.value) {
      urlParams.delete(param);
    } else {
      urlParams.set(param, this.value.toString());
    }
    window.history.replaceState(
      {},
      "",
      `${window.location.pathname}?${urlParams.toString()}`
    );
  }
}

customElements.define("filter-select", FilterSelect, { extends: "select" });
