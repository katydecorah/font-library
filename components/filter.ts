import { GeneratedData } from "./main-app";

export default function filter(
  state: {
    selectedSearch: string;
    selectedTag: string;
    selectedCategory: string;
    selectedSubset: string;
    selectedVariant: string;
    selectedVariable: boolean;
    sortBy: string;
    currentPage: number;
    pageSize: number;
  },
  generatedData: GeneratedData
): [number, GeneratedData] {
  const {
    selectedSearch,
    selectedTag,
    selectedCategory,
    selectedSubset,
    selectedVariant,
    selectedVariable,
    sortBy,
    currentPage,
    pageSize,
  } = state;
  let filteredData = generatedData;

  if (selectedSearch) {
    filteredData = filteredData.filter((row) =>
      row.family.toLowerCase().includes(selectedSearch.toLowerCase())
    );
  }

  if (selectedTag && selectedTag !== "need tags") {
    filteredData = filteredData.filter((row) => row.tags.includes(selectedTag));
  }
  if (selectedTag === "need tags") {
    filteredData = filteredData.filter((row) => row.tags.length === 0);
  }
  if (selectedCategory) {
    filteredData = filteredData.filter(
      (row) => row.category === selectedCategory
    );
  }
  if (selectedSubset) {
    filteredData = filteredData.filter((row) =>
      row.subsets.includes(selectedSubset)
    );
  }
  if (selectedVariant) {
    filteredData = filteredData.filter((row) =>
      row.variants.includes(selectedVariant)
    );
  }

  if (selectedVariable === true) {
    filteredData = filteredData.filter((row) => row.variable);
  }

  if (sortBy === "date") {
    filteredData = filteredData.sort(
      (a, b) =>
        new Date(b.lastModified).getTime() - new Date(a.lastModified).getTime()
    );
  }

  if (sortBy === "family") {
    filteredData = filteredData.sort((a, b) =>
      a.family.localeCompare(b.family)
    );
  }

  const paginated = filteredData.filter((row, index) => {
    const start = (currentPage - 1) * pageSize;
    const end = currentPage * pageSize;
    if (index >= start && index < end) return true;
  });

  return [filteredData.length, paginated];
}
