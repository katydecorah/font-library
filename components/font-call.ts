import { GeneratedData } from "./font-results";

export default function fontCall({
  variants,
  slug,
  selectedVariant,
  previewName,
}: {
  variants: GeneratedData[number]["variants"];
  slug: string;
  selectedVariant: string;
  previewName: string;
}): string {
  let fontCallString = slug;

  if (selectedVariant && selectedVariant !== "regular") {
    fontCallString += fontCallSelectedVariant(selectedVariant);
  }
  // if font doesn't have regular variant, add subset to font call
  if (!selectedVariant && !variants.includes("regular")) {
    fontCallString += fontCallVariant(variants);
  }

  fontCallString += `&text=${encodeURIComponent(previewName)}&display=swap`;

  return `https://fonts.googleapis.com/css2?family=${fontCallString}`;
}

function fontCallVariant(variants: GeneratedData[number]["variants"]): string {
  const [firstVariant] = variants;
  if (/\d+/g.test(firstVariant)) {
    return `:wght@${firstVariant}`;
  } else if (firstVariant.includes("italic")) {
    return `:ital@1`;
  }
}

function fontCallSelectedVariant(selectedVariant: string): string {
  const hasItalic = selectedVariant.includes("italic");
  const variantNumber = selectedVariant.match(/\d+/g); // get number from selectedVariant

  const variants = [];
  if (selectedVariant === "italic") {
    variants.push("ital@1");
  } else if (hasItalic) {
    variants.push("ital");
  }
  if (variantNumber && variantNumber[0]) {
    variants.push(`wght@${hasItalic ? "1," : ""}${variantNumber[0]}`);
  }
  return `:${variants.join(",")}`;
}
