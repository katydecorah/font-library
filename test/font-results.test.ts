import customEvent from "../components/custom-event";
import "./components";

describe("FontResults", () => {
  beforeAll(() => {
    window.HTMLElement.prototype.scrollIntoView = jest.fn();
  });

  test("renders correctly", () => {
    document.body.innerHTML = `<div id="content"><font-results sort-by="family" selected-category="" selected-subset="" selected-variant="" selected-tag="" selected-search="" selected-variable=""></font-results></div>`;
    const fontResults = document.querySelector("font-results");
    expect(fontResults).toMatchSnapshot();
  });

  test("fires a next-page custom event", () => {
    document.body.innerHTML = `<div id="content"><font-results sort-by="family" selected-category="" selected-subset="" selected-variant="" selected-tag="" selected-search="" selected-variable=""></font-results></div>`;
    const fontResults = document.querySelector("font-results");
    fontResults.dispatchEvent(customEvent("next-page"));
    const pagination = document.querySelector("pagination-buttons");
    expect(pagination.getAttribute("current-page")).toBe("2");
  });

  test("fires a prev-page custom event", () => {
    document.body.innerHTML = `<div id="content"><font-results sort-by="family" selected-category="" selected-subset="" selected-variant="" selected-tag="" selected-search="" selected-variable=""></font-results></div>`;
    const fontResults = document.querySelector("font-results");
    fontResults.dispatchEvent(customEvent("next-page"));
    fontResults.dispatchEvent(customEvent("previous-page"));
    const pagination = document.querySelector("pagination-buttons");
    expect(pagination.getAttribute("current-page")).toBe("1");
  });
});
