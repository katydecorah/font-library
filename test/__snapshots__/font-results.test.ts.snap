// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FontResults renders correctly 1`] = `
<font-results
  search=""
  selected-category=""
  selected-subset=""
  selected-tag=""
  selected-variable=""
  selected-variant=""
>
  <div
    id="search-status"
  >
    <search-status
      class="search-status"
      resultslength="1497"
      search=""
      selectedcategory=""
      selectedsubset=""
      selectedtag=""
      selectedvariable=""
      selectedvariant=""
    />
  </div>
  

  <div
    class="families"
    id="families"
  >
    <font-item
      font="{"family":"ABeeZee","variants":["regular","italic"],"subsets":["latin","latin-ext"],"category":"sans-serif","tags":["literacy","kids"],"lineNumber":2,"variable":false}"
      selectedsubset=""
      selectedtag=""
      selectedvariant=""
    />
    

    <font-item
      font="{"family":"Abel","variants":["regular"],"subsets":["latin"],"category":"sans-serif","tags":["condensed"],"lineNumber":3,"variable":false}"
      selectedsubset=""
      selectedtag=""
      selectedvariant=""
    />
    

    <font-item
      font="{"family":"Abhaya Libre","variants":["regular","500","600","700","800"],"subsets":["latin","latin-ext","sinhala"],"category":"serif","tags":[],"lineNumber":4,"variable":false}"
      selectedsubset=""
      selectedtag=""
      selectedvariant=""
    />
    

    <font-item
      font="{"family":"Aboreto","variants":["regular"],"subsets":["latin","latin-ext"],"category":"display","tags":["caps"],"lineNumber":5,"variable":false}"
      selectedsubset=""
      selectedtag=""
      selectedvariant=""
    />
    

    <font-item
      font="{"family":"Abril Fatface","variants":["regular"],"subsets":["latin","latin-ext"],"category":"display","tags":["heavy","elegant","didone","headlines"],"lineNumber":6,"variable":false}"
      selectedsubset=""
      selectedtag=""
      selectedvariant=""
    />
    

    <font-item
      font="{"family":"Abyssinica SIL","variants":["regular"],"subsets":["ethiopic","latin","latin-ext"],"category":"serif","tags":[],"lineNumber":7,"variable":false}"
      selectedsubset=""
      selectedtag=""
      selectedvariant=""
    />
    

    <font-item
      font="{"family":"Aclonica","variants":["regular"],"subsets":["latin"],"category":"sans-serif","tags":["modern","friendly"],"lineNumber":8,"variable":false}"
      selectedsubset=""
      selectedtag=""
      selectedvariant=""
    />
    

    <font-item
      font="{"family":"Acme","variants":["regular"],"subsets":["latin"],"category":"sans-serif","tags":["cartoon","comic","groovy","headlines"],"lineNumber":9,"variable":false}"
      selectedsubset=""
      selectedtag=""
      selectedvariant=""
    />
    

    <font-item
      font="{"family":"Actor","variants":["regular"],"subsets":["latin"],"category":"sans-serif","tags":["thin","legibility"],"lineNumber":10,"variable":false}"
      selectedsubset=""
      selectedtag=""
      selectedvariant=""
    />
    

    <font-item
      font="{"family":"Adamina","variants":["regular"],"subsets":["latin"],"category":"serif","tags":["compact","small sizes","transitional"],"lineNumber":11,"variable":false}"
      selectedsubset=""
      selectedtag=""
      selectedvariant=""
    />
  </div>
  

  <div
    id="pagination"
  >
    <pagination-buttons
      currentpage="1"
      pagesize="10"
      resultslength="1497"
    />
  </div>
</font-results>
`;
